File: tests/Tests.FSharp/Yaml/CrossVerifyTests.fs
module Zeta.Tests.FSharp.Yaml.CrossVerifyTests

// F# oracle (oracle #3) cross-verify + YamlDotNet differential.
//
// The fixture tests/cross-verification/yaml/vectors.json is JSON (NOT YAML -- do not use
// YAML to test YAML). Read it via System.Text.Json. For each vector run `readEvents`,
// assert the event list equals the fixture's `expected`, and serialize `{ id: events }` to
// fsharp-output.json in the same shape as ts-output.json so compare.ts can Bun.deepEquals
// TS == F# == Rust.
//
// Plus a differential [<Fact>]: drive YamlDotNet's own forward-only event-level parser
// (YamlDotNet.Core.Parser over a StringReader) and compare its event stream to OUR
// readEvents output -- structure + raw + style only (YamlDotNet does not resolve our
// core-schema `kind`, so kind is ignored; kind is OUR contract, already validated by
// vectors.json above).

open System
open System.Collections.Generic
open System.IO
open System.Text.Json
open Xunit
open Zeta.Core.FSharp.Yaml

// --- Repo-root walk (Zeta.sln sentinel; mirrors ZetaId CrossVerifyTests) -------------

/// Walk up from the test assembly looking for Zeta.sln (sentinel at repo root). .git is
/// unreliable (in a worktree it is a file, not a directory).
let private repoRoot () : string =
    let assembly = typeof<YamlEvent>.Assembly
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(assembly.Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then
        raise (InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location."))
    dir.FullName

let private yamlDir () : string =
    Path.Join(repoRoot (), "tests", "cross-verification", "yaml")

// --- Enum <-> contract string mapping (the cross-language names) ---------------------

let private kindName (k: ScalarKind) : string =
    match k with
    | ScalarKind.Null -> "Null"
    | ScalarKind.Bool -> "Bool"
    | ScalarKind.Int -> "Int"
    | ScalarKind.Float -> "Float"
    | ScalarKind.Str -> "Str"

let private styleName (s: ScalarStyle) : string =
    match s with
    | ScalarStyle.Plain -> "Plain"
    | ScalarStyle.SingleQuoted -> "SingleQuoted"
    | ScalarStyle.DoubleQuoted -> "DoubleQuoted"

let private kindOfName (name: string) : ScalarKind =
    match name with
    | "Null" -> ScalarKind.Null
    | "Bool" -> ScalarKind.Bool
    | "Int" -> ScalarKind.Int
    | "Float" -> ScalarKind.Float
    | "Str" -> ScalarKind.Str
    | other -> raise (InvalidOperationException(sprintf "Unknown ScalarKind name: %s" other))

let private styleOfName (name: string) : ScalarStyle =
    match name with
    | "Plain" -> ScalarStyle.Plain
    | "SingleQuoted" -> ScalarStyle.SingleQuoted
    | "DoubleQuoted" -> ScalarStyle.DoubleQuoted
    | other -> raise (InvalidOperationException(sprintf "Unknown ScalarStyle name: %s" other))

let private eventTag (e: YamlEvent) : string =
    match e with
    | StreamStart -> "StreamStart"
    | StreamEnd -> "StreamEnd"
    | MappingStart -> "MappingStart"
    | MappingEnd -> "MappingEnd"
    | SequenceStart -> "SequenceStart"
    | SequenceEnd -> "SequenceEnd"
    | Scalar _ -> "Scalar"

// --- JSON event (de)serialization (shape matches ts-output.json) --------------------

/// One event as an ordered string->string dict: { e } for structural, or
/// { e, raw, kind, style } for a Scalar. Bun.deepEquals is key-order-insensitive so a
/// plain Dictionary serializes correctly.
let private eventToJsonObj (e: YamlEvent) : Dictionary<string, string> =
    let d = Dictionary<string, string>(StringComparer.Ordinal)
    d.["e"] <- eventTag e
    match e with
    | Scalar(raw, kind, style) ->
        d.["raw"] <- raw
        d.["kind"] <- kindName kind
        d.["style"] <- styleName style
    | _ -> ()
    d

/// Parse one fixture event object back into a YamlEvent (for the assert against `expected`).
let private eventFromJson (el: JsonElement) : YamlEvent =
    let tag = el.GetProperty("e").GetString()
    match tag with
    | "StreamStart" -> StreamStart
    | "StreamEnd" -> StreamEnd
    | "MappingStart" -> MappingStart
    | "MappingEnd" -> MappingEnd
    | "SequenceStart" -> SequenceStart
    | "SequenceEnd" -> SequenceEnd
    | "Scalar" ->
        let raw = el.GetProperty("raw").GetString()
        let kind = kindOfName (el.GetProperty("kind").GetString())
        let style = styleOfName (el.GetProperty("style").GetString())
        Scalar(raw, kind, style)
    | other -> raise (InvalidOperationException(sprintf "Unknown event tag: %s" other))

// --- Fixture model ------------------------------------------------------------------

type private Vector =
    { Id: string
      Yaml: string
      Expected: YamlEvent list }

let private loadVectors () : Vector list =
    let path = Path.Join(yamlDir (), "vectors.json")
    let text = File.ReadAllText(path)
    use doc = JsonDocument.Parse(text)
    let vectorsEl = doc.RootElement.GetProperty("vectors")
    [ for v in vectorsEl.EnumerateArray() ->
        { Id = v.GetProperty("id").GetString()
          Yaml = v.GetProperty("yaml").GetString()
          Expected = [ for ev in v.GetProperty("expected").EnumerateArray() -> eventFromJson ev ] } ]

// --- Cross-verify: events == expected, write fsharp-output.json ----------------------

[<Fact>]
let ``cross-verify ten yaml vectors match fixture expected`` () =
    let vectors = loadVectors ()
    // Preserve fixture order in the output map so the file is stable; compare.ts is
    // key-order-insensitive but stable output keeps diffs clean.
    let results = Dictionary<string, Dictionary<string, string> list>(StringComparer.Ordinal)
    let mutable mismatches = 0

    for v in vectors do
        let events =
            match readEvents v.Yaml with
            | Ok evs -> evs
            | Error f -> failwithf "vector %s declined unexpectedly: %A" v.Id f
        if events <> v.Expected then
            mismatches <- mismatches + 1
        results.[v.Id] <- events |> List.map eventToJsonObj

    // Serialize { id: events } to fsharp-output.json (System.Text.Json; same indent style as
    // the ZetaId precedent). UnsafeRelaxedJsonEscaping so a literal newline inside a Scalar
    // "raw" serializes as the JSON escape backslash-n (matching ts-output.json).
    let options =
        JsonSerializerOptions(
            WriteIndented = true,
            Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
        )
    // Normalize to pure LF: net10 System.Text.Json indents with the platform newline
    // (CRLF on Windows); the cross-verification fixtures are pure-LF, so strip CRs.
    let json = JsonSerializer.Serialize(results, options).Replace("\r\n", "\n")
    File.WriteAllText(Path.Join(yamlDir (), "fsharp-output.json"), json)

    Assert.Equal(0, mismatches)

// --- YamlDotNet differential --------------------------------------------------------

/// Map YamlDotNet's forward-only event stream to our structural shape for comparison. We
/// compare structural events (Mapping/Sequence Start/End) by tag, and Scalar by (raw, style)
/// -- IGNORING our `kind` (YamlDotNet does not resolve core-schema kinds the same way; kind
/// is OUR contract, already validated against vectors.json). YamlDotNet's StreamStart/End +
/// DocumentStart/End wrapper events are skipped; our StreamStart/End are skipped too, so both
/// sides reduce to the document body.
type private DiffEvent =
    | DMapStart
    | DMapEnd
    | DSeqStart
    | DSeqEnd
    | DScalar of raw: string * style: ScalarStyle

let private ourDiffStream (events: YamlEvent list) : DiffEvent list =
    events
    |> List.choose (fun e ->
        match e with
        | StreamStart | StreamEnd -> None
        | MappingStart -> Some DMapStart
        | MappingEnd -> Some DMapEnd
        | SequenceStart -> Some DSeqStart
        | SequenceEnd -> Some DSeqEnd
        | Scalar(raw, _, style) -> Some(DScalar(raw, style)))

/// Translate a YamlDotNet scalar's quote style to our ScalarStyle.
let private vendorStyle (s: YamlDotNet.Core.ScalarStyle) : ScalarStyle =
    match s with
    | YamlDotNet.Core.ScalarStyle.SingleQuoted -> ScalarStyle.SingleQuoted
    | YamlDotNet.Core.ScalarStyle.DoubleQuoted -> ScalarStyle.DoubleQuoted
    | _ -> ScalarStyle.Plain // Plain / Any / Literal / Folded -> Plain (subset is plain/quoted)

let private vendorDiffStream (yaml: string) : DiffEvent list =
    use reader = new StringReader(yaml)
    let parser = YamlDotNet.Core.Parser(reader)
    let acc = List<DiffEvent>()
    while parser.MoveNext() do
        match parser.Current with
        | :? YamlDotNet.Core.Events.MappingStart -> acc.Add(DMapStart)
        | :? YamlDotNet.Core.Events.MappingEnd -> acc.Add(DMapEnd)
        | :? YamlDotNet.Core.Events.SequenceStart -> acc.Add(DSeqStart)
        | :? YamlDotNet.Core.Events.SequenceEnd -> acc.Add(DSeqEnd)
        | :? YamlDotNet.Core.Events.Scalar as sc -> acc.Add(DScalar(sc.Value, vendorStyle sc.Style))
        // Skip wrapper events: StreamStart/End, DocumentStart/End, and anything else.
        | _ -> ()
    List.ofSeq acc

[<Fact>]
let ``yamldotnet differential agrees on all vectors (structure + raw + style)`` () =
    let vectors = loadVectors ()
    let mutable mismatches = 0
    for v in vectors do
        let ours =
            match readEvents v.Yaml with
            | Ok evs -> ourDiffStream evs
            | Error f -> failwithf "vector %s declined unexpectedly: %A" v.Id f
        let theirs = vendorDiffStream v.Yaml
        if ours <> theirs then
            mismatches <- mismatches + 1
    Assert.Equal(0, mismatches)

File: tests/Tests.CSharp/Yaml/CrossVerifyTests.cs
// C# oracle (oracle #4) cross-verify + YamlDotNet differential.
//
// The fixture tests/cross-verification/yaml/vectors.json is JSON (NOT YAML -- do not use
// YAML to test YAML). Read it via System.Text.Json. For each vector run ReadEvents, assert
// the event list equals the fixture's expected, and serialize { id: events } to
// cs-output.json in the same shape as ts-output.json so compare.ts can Bun.deepEquals
// TS == F# == Rust == C#.
//
// Plus a differential [Fact]: drive YamlDotNet's own forward-only event-level parser
// (YamlDotNet.Core.Parser over a StringReader) and compare its event stream to OUR
// ReadEvents output -- structure + raw + style only (YamlDotNet does not resolve our
// core-schema kind, so kind is ignored; kind is OUR contract, already validated by
// vectors.json above).

namespace Zeta.Tests.CSharp.Yaml;

using System.IO;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp.Yaml;

public class CrossVerifyTests
{
    // -------------------------------------------------------------------------
    // Repo-root walk (Zeta.sln sentinel; mirrors ZetaId CrossVerifyTests)
    // -------------------------------------------------------------------------

    private static string RepoRoot()
    {
        DirectoryInfo? dir = new DirectoryInfo(Path.GetDirectoryName(typeof(CrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
            dir = dir.Parent;
        return dir?.FullName
            ?? throw new InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location.");
    }

    private static string YamlDir() => Path.Join(RepoRoot(), "tests", "cross-verification", "yaml");

    // -------------------------------------------------------------------------
    // Enum <-> contract string mapping (the cross-language names)
    // -------------------------------------------------------------------------

    private static string KindName(ScalarKind k) => k switch
    {
        ScalarKind.Null  => "Null",
        ScalarKind.Bool  => "Bool",
        ScalarKind.Int   => "Int",
        ScalarKind.Float => "Float",
        _                => "Str",
    };

    private static string StyleName(ScalarStyle s) => s switch
    {
        ScalarStyle.SingleQuoted => "SingleQuoted",
        ScalarStyle.DoubleQuoted => "DoubleQuoted",
        _                        => "Plain",
    };

    private static ScalarKind KindOfName(string name) => name switch
    {
        "Null"  => ScalarKind.Null,
        "Bool"  => ScalarKind.Bool,
        "Int"   => ScalarKind.Int,
        "Float" => ScalarKind.Float,
        "Str"   => ScalarKind.Str,
        _       => throw new InvalidOperationException($"Unknown ScalarKind name: {name}"),
    };

    private static ScalarStyle StyleOfName(string name) => name switch
    {
        "SingleQuoted" => ScalarStyle.SingleQuoted,
        "DoubleQuoted" => ScalarStyle.DoubleQuoted,
        "Plain"        => ScalarStyle.Plain,
        _              => throw new InvalidOperationException($"Unknown ScalarStyle name: {name}"),
    };

    // -------------------------------------------------------------------------
    // JSON event (de)serialization (shape matches ts-output.json)
    // -------------------------------------------------------------------------

    private static YamlEvent EventFromJson(JsonElement el)
    {
        string tag = el.GetProperty("e").GetString()!;
        return tag switch
        {
            "StreamStart"   => new YamlEvent.StreamStart(),
            "StreamEnd"     => new YamlEvent.StreamEnd(),
            "MappingStart"  => new YamlEvent.MappingStart(),
            "MappingEnd"    => new YamlEvent.MappingEnd(),
            "SequenceStart" => new YamlEvent.SequenceStart(),
            "SequenceEnd"   => new YamlEvent.SequenceEnd(),
            "Scalar" =>
                new YamlEvent.Scalar(
                    el.GetProperty("raw").GetString()!,
                    KindOfName(el.GetProperty("kind").GetString()!),
                    StyleOfName(el.GetProperty("style").GetString()!)),
            _ => throw new InvalidOperationException($"Unknown event tag: {tag}"),
        };
    }

    // Serialize one event to the cross-verify JSON shape.
    // Uses manual StringBuilder to avoid JsonSerializer overhead and control key order.
    private static string EventToJson(YamlEvent e)
    {
        if (e is YamlEvent.Scalar sc)
        {
            // { "e": "Scalar", "raw": "...", "kind": "...", "style": "..." }
            // Use JsonSerializer to handle escaping of raw and kind/style strings.
            string rawJson  = JsonSerializer.Serialize(sc.Raw);
            string kindJson = JsonSerializer.Serialize(KindName(sc.Kind));
            string styleJson = JsonSerializer.Serialize(StyleName(sc.Style));
            return $"{{\"e\":\"Scalar\",\"raw\":{rawJson},\"kind\":{kindJson},\"style\":{styleJson}}}";
        }
        string tag = e switch
        {
            YamlEvent.StreamStart   => "StreamStart",
            YamlEvent.StreamEnd     => "StreamEnd",
            YamlEvent.MappingStart  => "MappingStart",
            YamlEvent.MappingEnd    => "MappingEnd",
            YamlEvent.SequenceStart => "SequenceStart",
            YamlEvent.SequenceEnd   => "SequenceEnd",
            _                       => throw new InvalidOperationException($"Unknown event: {e}"),
        };
        return $"{{\"e\":\"{tag}\"}}";
    }

    // -------------------------------------------------------------------------
    // Fixture model
    // -------------------------------------------------------------------------

    private sealed record Vector(string Id, string Yaml, List<YamlEvent> Expected);

    private static List<Vector> LoadVectors()
    {
        string path = Path.Join(YamlDir(), "vectors.json");
        string text = File.ReadAllText(path);
        using JsonDocument doc = JsonDocument.Parse(text);
        var result = new List<Vector>();
        foreach (JsonElement v in doc.RootElement.GetProperty("vectors").EnumerateArray())
        {
            string id   = v.GetProperty("id").GetString()!;
            string yaml = v.GetProperty("yaml").GetString()!;
            var expected = new List<YamlEvent>();
            foreach (JsonElement evEl in v.GetProperty("expected").EnumerateArray())
                expected.Add(EventFromJson(evEl));
            result.Add(new Vector(id, yaml, expected));
        }
        return result;
    }

    // -------------------------------------------------------------------------
    // Cross-verify: events == expected, write cs-output.json
    // -------------------------------------------------------------------------

    [Fact]
    public void CrossVerifyTenYamlVectorsMatchFixtureExpected()
    {
        List<Vector> vectors = LoadVectors();
        var sb = new StringBuilder();
        sb.Append('{');
        bool firstId = true;
        int mismatches = 0;

        foreach (Vector v in vectors)
        {
            ReadResult r = YamlReader.ReadEvents(v.Yaml);
            Assert.True(r.Ok, $"vector {v.Id} declined unexpectedly: feedback={r.Feedback}");
            IReadOnlyList<YamlEvent> actual = r.Events!;

            // Assert events equal the fixture's expected list.
            bool match = actual.Count == v.Expected.Count;
            if (match)
                for (int i = 0; i < actual.Count; i++)
                    if (!actual[i].Equals(v.Expected[i])) { match = false; break; }
            if (!match) mismatches++;

            // Build the JSON for this vector's entry.
            if (!firstId) sb.Append(',');
            firstId = false;
            sb.Append('\n');
            sb.Append("  ");
            sb.Append(JsonSerializer.Serialize(v.Id));
            sb.Append(": [\n");
            bool firstEv = true;
            foreach (YamlEvent ev in actual)
            {
                if (!firstEv) sb.Append(",\n");
                firstEv = false;
                // Indent each event JSON with 4 spaces for readability (matches ts-output.json indent).
                string evJson = EventToJson(ev);
                sb.Append("    ");
                sb.Append(evJson);
            }
            sb.Append('\n');
            sb.Append("  ]");
        }

        sb.Append('\n');
        sb.Append('}');
        sb.Append('\n');

        // Normalise to pure LF (the cross-verification fixtures are pure-LF).
        string json = sb.ToString().Replace("\r\n", "\n", StringComparison.Ordinal);
        File.WriteAllText(Path.Join(YamlDir(), "cs-output.json"), json);

        Assert.Equal(0, mismatches);
    }

    // The ts-output.json shape uses WriteIndented=true with 2-space indentation per key.
    // We need to match it closely so compare.ts deep-equals pass. Let us just serialize
    // with a matching style using JsonSerializerOptions directly.
    // Actually: compare.ts uses Bun.deepEquals on parsed JSON -- key/value equality only,
    // not text equality. So we just need valid JSON with the right structure.
    // EventToJson already produces compact JSON per event; the file wrapper above wraps it.
    // We produce minimal indented-enough output. compare.ts parses the JSON anyway.


    // -------------------------------------------------------------------------
    // YamlDotNet differential
    // -------------------------------------------------------------------------

    private enum DiffTag { MapStart, MapEnd, SeqStart, SeqEnd, Scalar }

    private readonly record struct DiffEvent(DiffTag Tag, string? Raw, ScalarStyle Style);

    private static List<DiffEvent> OurDiffStream(IReadOnlyList<YamlEvent> events)
    {
        var result = new List<DiffEvent>();
        foreach (YamlEvent ev in events)
        {
            switch (ev)
            {
                case YamlEvent.MappingStart:  result.Add(new DiffEvent(DiffTag.MapStart, null, default)); break;
                case YamlEvent.MappingEnd:    result.Add(new DiffEvent(DiffTag.MapEnd,   null, default)); break;
                case YamlEvent.SequenceStart: result.Add(new DiffEvent(DiffTag.SeqStart, null, default)); break;
                case YamlEvent.SequenceEnd:   result.Add(new DiffEvent(DiffTag.SeqEnd,   null, default)); break;
                case YamlEvent.Scalar sc:     result.Add(new DiffEvent(DiffTag.Scalar, sc.Raw, sc.Style)); break;
                // StreamStart / StreamEnd skipped
                default: break;
            }
        }
        return result;
    }

    private static ScalarStyle VendorStyle(YamlDotNet.Core.ScalarStyle s) => s switch
    {
        YamlDotNet.Core.ScalarStyle.SingleQuoted => ScalarStyle.SingleQuoted,
        YamlDotNet.Core.ScalarStyle.DoubleQuoted => ScalarStyle.DoubleQuoted,
        _                                         => ScalarStyle.Plain,
    };

    private static List<DiffEvent> VendorDiffStream(string yaml)
    {
        using var reader = new StringReader(yaml);
        var parser = new YamlDotNet.Core.Parser(reader);
        var acc = new List<DiffEvent>();
        while (parser.MoveNext())
        {
            switch (parser.Current)
            {
                case YamlDotNet.Core.Events.MappingStart:
                    acc.Add(new DiffEvent(DiffTag.MapStart, null, default)); break;
                case YamlDotNet.Core.Events.MappingEnd:
                    acc.Add(new DiffEvent(DiffTag.MapEnd,   null, default)); break;
                case YamlDotNet.Core.Events.SequenceStart:
                    acc.Add(new DiffEvent(DiffTag.SeqStart, null, default)); break;
                case YamlDotNet.Core.Events.SequenceEnd:
                    acc.Add(new DiffEvent(DiffTag.SeqEnd,   null, default)); break;
                case YamlDotNet.Core.Events.Scalar sc:
                    acc.Add(new DiffEvent(DiffTag.Scalar, sc.Value, VendorStyle(sc.Style))); break;
                // Skip StreamStart/End, DocumentStart/End, and anything else.
                default: break;
            }
        }
        return acc;
    }

    [Fact]
    public void YamlDotNetDifferentialAgreesOnAllVectors()
    {
        List<Vector> vectors = LoadVectors();
        int mismatches = 0;
        foreach (Vector v in vectors)
        {
            ReadResult r = YamlReader.ReadEvents(v.Yaml);
            Assert.True(r.Ok, $"vector {v.Id} declined unexpectedly: feedback={r.Feedback}");

            List<DiffEvent> ours   = OurDiffStream(r.Events!);
            List<DiffEvent> theirs = VendorDiffStream(v.Yaml);

            if (ours.Count != theirs.Count) { mismatches++; continue; }
            for (int i = 0; i < ours.Count; i++)
                if (ours[i] != theirs[i]) { mismatches++; break; }
        }
        Assert.Equal(0, mismatches);
    }
}
